### PR TITLE
plugin to cache the exif data of images

### PR DESCRIPTION
**Pull request by t-animal**
Decreases the time needed to build large galleries (e.g.: 25k images in 2.5s instead of 30s)

This plugin allows extended caching, which is useful for large galleries. Once a gallery has
been built it caches the exif-data of the contained images in the gallery target folder.
Before the next run it restores them so that the image does not have to be parsed again. For
large galleries this can speed up the creation of index files dramatically.

### CODE REVIEW

File: sigal/plugins/extended_caching.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    """Restores the exif data cache from the cache file"""
+    cachePath = os.path.join(gallery.settings["destination"], ".exif_cache")
+    try:
+        if os.path.exists(cachePath):
+            with open(cachePath, "rb") as cacheFile:
+                gallery.exifCache = pickle.load(cacheFile)
+                logger.debug("Loaded cache with %d entries", len(gallery.exifCache))
+        else:
+            gallery.exifCache = {}
+    except Exception as e:
```
Comment (saimn) [inline sigal/plugins/extended_caching.py]: Maybe move this try/except around the `pickle.load` line ?

File: sigal/plugins/extended_caching.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        logger.warn("Could not load cache: %s", e)
+        gallery.exifCache = {}
+
+def save_cache(gallery):
+    """Stores the exif data of all images in the gallery"""
+
+    if hasattr(gallery, "exifCache"):
+        cache = gallery.exifCache
+    else:
+        cache = gallery.exifCache = {}
```
Comment (saimn) [inline sigal/plugins/extended_caching.py]: Could be simplified with ``getattr(gallery, "exifCache", {})`` and always use `cache` below.

**Comment by saimn** [comment]
@t-animal - Can you rebase on master to get rid of the conflict on `tests/sample/sigal.conf.py` ?

File: sigal/plugins/extended_caching.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+logger = logging.getLogger(__name__)
+
+def load_exif(album):
+    """Loads the exif data of all images in an album from cache"""
+    if not hasattr(album.gallery, "exifCache"):
+        _restore_cache(album.gallery)
+    cache = album.gallery.exifCache
+
+    for media in album.medias:
+        if media.type == "image":
```
Comment (saimn) [inline sigal/plugins/extended_caching.py]: Forgot this, you can use `album.images`.

**Comment by saimn** [comment]
It would also be useful to add your plugins to the related documentation page, with a quick description: http://sigal.saimon.org/en/latest/plugins.html

**Comment by krylovsk** [comment]
This is neat, I've been experiencing long rebuild times with my moderate (~5k) gallery.
Documentation would be welcome.

File: sigal/plugins/extended_caching.py (hunk 1 above)
Comment (t-animal) [inline sigal/plugins/extended_caching.py]: This is intended, as the file might exist but not be readable for the user.

File: sigal/plugins/extended_caching.py (hunk 2 above)
Comment (t-animal) [inline sigal/plugins/extended_caching.py]: mhm, I would have to change the test cases, as they assume gallery.exifCache exists...

File: sigal/plugins/extended_caching.py (hunk 3 above)
Comment (t-animal) [inline sigal/plugins/extended_caching.py]: fixed

**Comment by t-animal** [comment]
@krylovsk simply include the plugin in your config (plugins=['sigal.plugins.extended_caching']) and after you've build the gallery the next time, a cache-file will be put in your gallery from where the cached data will be loaded upon the build after that.

**Comment by saimn** [comment]
Merging, thanks @t-animal !

**Pull request merged**